lib/cusparse/interfaces.jl

```diff
@@ -75,74 +76,143 @@ for (taga, untaga) in tag_wrappers, (wrapa, transa, unwrapa) in op_wrappers
+Base.:(-)(A::CuSparseMatrixCSR, B::Transpose{T,<:CuSparseMatrixCSR}) where {T} = A - _sptranspose(parent(B))
+Base.:(+)(A::Transpose{T,<:CuSparseMatrixCSR}, B::CuSparseMatrixCSR) where {T} = _sptranspose(parent(A)) + B
+Base.:(-)(A::Transpose{T,<:CuSparseMatrixCSR}, B::CuSparseMatrixCSR) where {T} = _sptranspose(parent(A)) - B
+Base.:(+)(A::Transpose{T,<:CuSparseMatrixCSR}, B::Transpose{T,<:CuSparseMatrixCSR}) where {T} = 
+    _sptranspose(parent(A)) + _sptranspose(parent(B))
+Base.:(-)(A::Transpose{T,<:CuSparseMatrixCSR}, B::Transpose{T,<:CuSparseMatrixCSR}) where {T} = 
+    _sptranspose(parent(A)) - _sptranspose(parent(B))
+
+Base.:(+)(A::CuSparseMatrixCSR, B::CuSparseMatrix) = A + CuSparseMatrixCSR(B)
+Base.:(-)(A::CuSparseMatrixCSR, B::CuSparseMatrix) = A - CuSparseMatrixCSR(B)
+Base.:(+)(A::CuSparseMatrix, B::CuSparseMatrixCSR) = CuSparseMatrixCSR(A) + B
+Base.:(-)(A::CuSparseMatrix, B::CuSparseMatrixCSR) = CuSparseMatrixCSR(A) - B
+
+Base.:(+)(A::Union{CuSparseMatrixCSC{T}, Transpose{T,<:CuSparseMatrixCSC}, Adjoint{T,<:CuSparseMatrixCSC}}, 
+          B::Union{CuSparseMatrixCSC{T}, Transpose{T,<:CuSparseMatrixCSC}, Adjoint{T,<:CuSparseMatrixCSC}}) where {T} =
+    CuSparseMatrixCSC(CuSparseMatrixCSR(A) + CuSparseMatrixCSR(B))
+Base.:(-)(A::Union{CuSparseMatrixCSC{T}, Transpose{T,<:CuSparseMatrixCSC}, Adjoint{T,<:CuSparseMatrixCSC}}, 
+          B::Union{CuSparseMatrixCSC{T}, Transpose{T,<:CuSparseMatrixCSC}, Adjoint{T,<:CuSparseMatrixCSC}}) where {T} =
+    CuSparseMatrixCSC(CuSparseMatrixCSR(A) - CuSparseMatrixCSR(B))
+
+Base.:(+)(A::Union{CuSparseMatrixCOO{T}, Transpose{T,<:CuSparseMatrixCOO}, Adjoint{T,<:CuSparseMatrixCOO}}, 
+          B::Union{CuSparseMatrixCOO{T}, Transpose{T,<:CuSparseMatrixCOO}, Adjoint{T,<:CuSparseMatrixCOO}}) where {T} =
+    CuSparseMatrixCOO(CuSparseMatrixCSR(A) + CuSparseMatrixCSR(B))
+Base.:(-)(A::Union{CuSparseMatrixCOO{T}, Transpose{T,<:CuSparseMatrixCOO}, Adjoint{T,<:CuSparseMatrixCOO}}, 
+          B::Union{CuSparseMatrixCOO{T}, Transpose{T,<:CuSparseMatrixCOO}, Adjoint{T,<:CuSparseMatrixCOO}}) where {T} =
+    CuSparseMatrixCOO(CuSparseMatrixCSR(A) - CuSparseMatrixCSR(B))
+
+function Base.reshape(A::CuSparseMatrixCOO{T,M}, dims::NTuple{N,Int}) where {T,N,M}
+    nrows, ncols = size(A)
+    flat_indices = nrows .* (A.colInd .- 1) .+ A.rowInd .- 1
+    new_col, new_row = div.(flat_indices, dims[1]) .+ 1, rem.(flat_indices, dims[1]) .+ 1
+    sparse(new_row, new_col, A.nzVal, dims[1], length(dims) == 1 ? 1 : dims[2], fmt = :coo)
 end
 
-function Base.:(-)(A::CuSparseMatrixCSR, B::Transpose{T,<:CuSparseMatrixCSR}) where {T}
-    cscB = CuSparseMatrixCSC(B.parent)
-    transB = CuSparseMatrixCSR(cscB.colPtr, cscB.rowVal, cscB.nzVal, size(cscB))
-    return geam(one(T), A, -one(T), transB, 'O')
+function LinearAlgebra.mul!(Y::CuSparseMatrixCSR{T,M}, A::CuSparseMatrixCSR{T,M}, B::CuSparseMatrixCSR{T,M}, alpha::Number, beta::Number) where {T,M}
```

Comment:
Could you format these a little? We generally try to stick to the 92 column limit, if possible.

Reply:
> CUSPARSE has a routine for `droptol!` (https://docs.nvidia.com/cuda/cusparse/index.html#pruneCsr2csr)

Cool! How to implement it? I saw that there is no julia function.

Reply:
> Could you format these a little? We generally try to stick to the 92 column limit, if possible.

Let me know if now it is ok.

Reply:
That's better, thanks. I'll have another formatting pass right before merging (we really ought to automate this).